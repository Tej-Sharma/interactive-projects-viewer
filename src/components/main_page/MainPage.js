import React, { Component } from "react";
import MapDisplay from "./page_components/MapDisplay";
// React bootstrap componenets
import Form from "react-bootstrap/Form";
// Retrieve the projects data from the JSON file
import projectsData from "../../data/projects";
// For posting requests to the backend server
import axios from "axios";

import Roll from "react-reveal/Roll";

export class MainPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      // Track the parameters that are checked in a dictionary to make updating simple
      checkboxes: {
        strategyChecked: true,
        engineeringChecked: true,
        infrastructureChecked: true,
        designChecked: true,
      },
      rawProjectsData: [], // The raw data retrieved from MongoDB will be stored here
      // Store the data in a dictionary where the key represents the
      projectsByCountry: {},
    };

    // Bind the function (so that it can call methods related to this class)
    this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
  }

  /**
   * Use this to load in the data from MongoDB and then parse into a more efficient, usable format
   */
  componentDidMount() {
    // IMPORTANT: this does not need to be called, but for future reference, will be left here
    // Only needed to be called once to convert and push JSON to MongoDB, but that's done now
    //this.pushDataToDB();

    // Load in the data
    this.loadDataFromDB();

    // Initially, all 3 filters are set so a mix of all types of services will
    // be loaded in
    this.filterData();
  }

  /**
   * Use a get request to retrieve data
   */
  loadDataFromDB() {
    axios
      .get("http://localhost:5000/")
      .then((res) => {
        this.setState({ rawProjectsData: res.data });
        this.filterData();
      })
      .catch((err) => console.log("Error posting data"));
  }

  // Load DB for integration TESTING only, only used in MainPage.test.js to test
  // frontend (axios)/backend(express) integration
  testLoadDataFromDB() {
    return new Promise((resolve, reject) => {
      axios({
        method: "GET",
      }).then(({ status, data }) => {
        return status;
      });
    });
  }

  /**
   * Post a request to MongoDB to add all of the projects
   * Load data from the JSON and then convert it to a MongoDB model and then push to the DBs
   * NOTE: do not call this method more than once as this will lead to duplicate
   * data being added (it is just a one-time helper method, but I've included it here
   * to show how I did it orignally)
   */
  pushDataToDB() {
    console.log(projectsData.projects);
    projectsData.projects.forEach((project) => {
      axios
        .post("http://localhost:5000/add", project)
        .then((res) => console.log(res.data))
        .catch((err) => console.log("Error posting data"));
    });
  }

  /**
   * Wrote an algorithm to convert the array of data into a dictionary where the
   * key represents the country name and the values are all the projects for that country
   * This helps make looking up projects when the user clicks on a country much more efficient
   * as opposed to searching the list, complexity O(n), each time
   * In addition, this also handles filtering the data based on the user input to choose
   * what type of projects to display via the checkboxes
   */
  filterData() {
    // Get the current user selected filters based on the checkboxes
    let filters = [];

    Object.keys(this.state.checkboxes).forEach((key) => {
      if (this.state.checkboxes[key]) {
        filters = filters.concat(key.replace("Checked", ""));
      }
    });

    // The final parsed data that will be stored
    // Through the iterations below, this will be filled
    let parsedData = {};

    // Parse the data so that it is in the format: 'country_name': [all_projects]
    this.state.rawProjectsData.forEach((project) => {
      // Check if an entry on that key already exists
      if (parsedData[project.country]) {
        // FILTER the data by making sure the type of service matches the checkboxes
        // according to what the user wants to filter
        if (
          parsedData[project.country].length <= 2 &&
          filters.includes(project.services.toLowerCase())
        ) {
          parsedData[project.country] = parsedData[project.country].concat(
            project
          );
        }
      } else {
        // If an entry doesn't exist, then need to set up a list with its first element
        if (filters.includes(project.services.toLowerCase())) {
          parsedData[project.country] = [project];
        }
      }
    });

    // Finally store the data
    this.setState({ projectsByCountry: parsedData });
  }

  /**
   * Update the checked parameters
   * @param {*} evt - checked object
   */
  handleCheckboxClick(evt) {
    // Note: not conventional in React, but we are following this up with 'forceUpdate'
    // (does not matter as shouldComponentUpdate() is not needed here)
    this.state.checkboxes[evt.target.name] = !this.state.checkboxes[
      evt.target.name
    ];
    this.forceUpdate();
    // Then, filter the data
    this.filterData();
  }

  render() {
    return (
      <div>
        <Roll left>
          <div
            key="inline-checkbox"
            className="mb-3"
            style={{ margin: "8px 12px" }}
          >
            <Form.Check
              inline
              label="Strategy"
              name="strategyChecked"
              type="checkbox"
              id="strategyCheckedBox"
              checked={this.state.checkboxes["strategyChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Engineering"
              name="engineeringChecked"
              type="checkbox"
              id="engineeringCheckedBox"
              checked={this.state.checkboxes["engineeringChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Infrastructure"
              name="infrastructureChecked"
              type="checkbox"
              id="infrastructureCheckedBox"
              checked={this.state.checkboxes["infrastructureChecked"]}
              onChange={this.handleCheckboxClick}
            />
            <Form.Check
              inline
              label="Design"
              name="designChecked"
              type="checkbox"
              id="designCheckedBox"
              checked={this.state.checkboxes["designChecked"]}
              onChange={this.handleCheckboxClick}
            />
          </div>
        </Roll>

        <MapDisplay projectsByCountry={this.state.projectsByCountry} />
      </div>
    );
  }
}

export default MainPage;
