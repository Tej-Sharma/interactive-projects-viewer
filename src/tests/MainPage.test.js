import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import MainPage from "../components/main_page/MainPage";

// Main Page's unit testing
describe("MainPage tests", () => {
  // First, snapshot test the UI to ensure during development,
  // the UI does not mistakingly get changed with erraneous features
  test("MainPage renders correctly", () => {
    const tree = renderer.create(<MainPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Checkboxes
  test("handleCheckboxClick(): changing state correctly on user input", () => {
    const wrapper = shallow(<MainPage />);

    // First, get the checkboxes and simulate a change click on them
    const checkbox1 = wrapper.find("#strategyCheckedBox").first();
    const checkbox2 = wrapper.find("#engineeringCheckedBox").first();
    const checkbox3 = wrapper.find("#infrastructureCheckedBox").first();
    const checkbox4 = wrapper.find("#designCheckedBox").first();

    // Turn them off from true to false
    checkbox1.simulate("change", { target: { name: "strategyChecked" } });
    checkbox2.simulate("change", { target: { name: "engineeringChecked" } });
    checkbox3.simulate("change", { target: { name: "infrastructureChecked" } });
    checkbox4.simulate("change", { target: { name: "designChecked" } });

    // Test this case
    expect(wrapper.state("checkboxes")["strategyChecked"]).toEqual(false);
    expect(wrapper.state("checkboxes")["engineeringChecked"]).toEqual(false);
    expect(wrapper.state("checkboxes")["infrastructureChecked"]).toEqual(false);
    expect(wrapper.state("checkboxes")["designChecked"]).toEqual(false);

    // Turn them back on by clicking them again
    checkbox1.simulate("change", { target: { name: "strategyChecked" } });
    checkbox2.simulate("change", { target: { name: "engineeringChecked" } });
    checkbox3.simulate("change", { target: { name: "infrastructureChecked" } });
    checkbox4.simulate("change", { target: { name: "designChecked" } });

    // Test that they are now true
    expect(wrapper.state("checkboxes")["strategyChecked"]).toEqual(true);
    expect(wrapper.state("checkboxes")["engineeringChecked"]).toEqual(true);
    expect(wrapper.state("checkboxes")["infrastructureChecked"]).toEqual(true);
    expect(wrapper.state("checkboxes")["designChecked"]).toEqual(true);
  });

  // Filtering data works: correct filters and less than <3
  test("Data is filtered properly based on useri input", () => {
    const wrapper = shallow(<MainPage />);

    // Get the 'strategy' checkbox and simulate a click on it
    const checkbox1 = wrapper.find("#strategyCheckedBox").first();
    checkbox1.simulate("change", { target: { name: "strategyChecked" } });

    // Check filter data worked by checking the state
    const projectsByCountry = wrapper.state("projectsByCountry");

    Object.keys(projectsByCountry).forEach((key) => {
      // TEST that the length is less than 4 (so a maximum of 3 elements)
      expect(projectsByCountry[key].length).toBeLessThan(4);

      // Iterate and check that no 'Strategy' service is there
      // so filter is working correctly
      projectsByCountry[key].forEach((project) => {
        expect(project.services.toLowerCase()).not.toBe("strategy");
      });
    });

    // Just to be sure, test this once again by checking off 'Engineering'
    const checkbox2 = wrapper.find("#engineeringCheckedBox").first();
    checkbox2.simulate("change", { target: { name: "engineeringChecked" } });

    const projectsByCountrySecond = wrapper.state("projectsByCountry");

    Object.keys(projectsByCountrySecond).forEach((key) => {
      // TEST that the length is less than 4 (so a maximum of 3 elements)
      expect(projectsByCountrySecond[key].length).toBeLessThan(4);

      // Iterate and check that no 'Engineering' service is there
      // so filter is working correctly
      projectsByCountrySecond[key].forEach((project) => {
        // No engineering projects
        expect(project.services.toLowerCase()).not.toBe("engineering");
        // No strateegy projects as well
        expect(project.services.toLowerCase()).not.toBe("strategy");
      });
    });

    // Now if it works for these two and since state is correctly handled,
    // we can deduce that user input and filtering works properly

    // Finally, check if the correct props from filtering are passed
    expect(wrapper.find("MapDisplay").props().projectsByCountry).toEqual(
      projectsByCountrySecond
    );
  });
});
