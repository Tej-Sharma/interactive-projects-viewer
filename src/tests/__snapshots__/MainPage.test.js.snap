// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage tests MainPage renders correctly 1`] = `
<div>
  <div
    className="mb-3"
    style={
      Object {
        "margin": "8px 12px",
      }
    }
  >
    <div
      className="form-check form-check-inline"
    >
      <input
        checked={true}
        className="form-check-input"
        disabled={false}
        id="strategyCheckedBox"
        name="strategyChecked"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="strategyCheckedBox"
        title=""
      >
        Strategy
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        checked={true}
        className="form-check-input"
        disabled={false}
        id="engineeringCheckedBox"
        name="engineeringChecked"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="engineeringCheckedBox"
        title=""
      >
        Engineering
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        checked={true}
        className="form-check-input"
        disabled={false}
        id="infrastructureCheckedBox"
        name="infrastructureChecked"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="infrastructureCheckedBox"
        title=""
      >
        Infrastructure
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <input
        checked={true}
        className="form-check-input"
        disabled={false}
        id="designCheckedBox"
        name="designChecked"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="designCheckedBox"
        title=""
      >
        Design
      </label>
    </div>
  </div>
  <div
    className="map-parent-container"
  >
    <div
      className="map-parent-content"
    >
      <svg
        className="rsm-svg map"
        viewBox="0 0 800 600"
      >
        <g
          className="rsm-geographies "
        />
      </svg>
    </div>
  </div>
</div>
`;
